src/python/pants/backend/jvm/tasks/coverage/factory.py

```diff
@@ -0,0 +1,35 @@
+# coding=utf-8
+# Copyright 2014 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import (absolute_import, division, generators, nested_scopes, print_function,
+                        unicode_literals, with_statement)
+
+from pants.backend.jvm.tasks.coverage.base import BaseCoverage, NoCoverage
+from pants.backend.jvm.tasks.coverage.cobertura import Cobertura, CoberturaTaskSettings
+from pants.backend.jvm.tasks.coverage.jacoco import Jacoco, JacocoTaskSettings
+
+
+class CoverageFactory(object):
+  """A factory for creating code coverage classes and registering their options."""
+
+  @staticmethod
+  def from_task(task, output_dir, all_targets, execute_java):
+    coverage = NoCoverage()
+    options = task.get_options()
+
+    if options.coverage or options.coverage_processor or options.is_flagged('coverage_open'):
+      if options.coverage_processor == 'cobertura':
+        settings = CoberturaTaskSettings.from_task(task, workdir=output_dir)
+        coverage = Cobertura(settings, all_targets, execute_java)
+    elif options.coverage-processor == 'jacoco':
+        settings = JacocoTaskSettings.from_task(task, workdir=output_dir)
+        coverage = Jacoco(settings, all_targets, execute_java)
+
+    return coverage
+
+  @staticmethod
+  def register_options(register, register_jvm_tool):
```

Comment:
Rather than a factory for instantiating all three of these, I think that the clearest implementation at this point would be to have this be a graph of `Subsystems`. Subsystems are like `Tasks`, in that they can have options. They can also have dependencies on other Subsystems. Finally, Tasks can have dependencies on Subsystems.

Concretely: `JunitRun` already has a dependency on two subsystems:
https://github.com/pantsbuild/pants/blob/master/src/python/pants/backend/jvm/tasks/junit_run.py#L193-L195
...and I'm suggesting that Jacoco and Cobertura become Subsystems (ie, subclassing Subsystem), and that JunitRun should depend on them. Then, the "use a coverage-processor" option would stay in JunitRun, but would select which Subsystem to use.